### PR TITLE
Define an explicit device resource as DeviceV1

### PR DESCRIPTION
**Pull request by codingllama**
Declare an explicit device resource representation as the DeviceV1 proto, under api/types.

The intent is to make the Terraform integration easier by using the DeviceV1 proto directly.

https://github.com/gravitational/teleport.e/issues/939

### CODE REVIEW

File: api/types/device.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+	"github.com/google/uuid"
+	"github.com/gravitational/trace"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	devicepb "github.com/gravitational/teleport/api/gen/proto/go/teleport/devicetrust/v1"
+)
+
+// CheckAndSetDefaults checks DeviceV1 fields to catch simple errors, and sets
+// default values for all fields with defaults.
+func (d *DeviceV1) CheckAndSetDefaults() error {
```
Comment (codingllama) [inline api/types/device.go:30]: Moved from tool/tctl/common/device/resource.go, made public so it takes over from the its namesake from ResourceHeader.

File: api/types/device.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+	}
+	if _, err := ResourceEnrollStatusFromString(d.Spec.EnrollStatus); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// UnmarshalDevice unmarshals a DeviceV1 resource and runs CheckAndSetDefaults.
+func UnmarshalDevice(raw []byte) (*DeviceV1, error) {
```
Comment (codingllama) [inline api/types/device.go:72]: Moved from tool/tctl/common/device/resource.go.

File: api/types/device.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+	dev := &DeviceV1{}
+	if err := json.Unmarshal(raw, dev); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return dev, trace.Wrap(dev.CheckAndSetDefaults())
+}
+
+// DeviceFromResource converts a resource DeviceV1 to an API devicepb.Device.
+func DeviceFromResource(res *DeviceV1) (*devicepb.Device, error) {
```
Comment (codingllama) [inline api/types/device.go:81]: Moved from tool/tctl/common/device/resource.go, previously `resourceToProto`.

File: api/types/device.go [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+		UpdateTime:    toTimePB(res.Spec.UpdateTime),
+		EnrollStatus:  enrollStatus,
+		Credential:    cred,
+		CollectedData: collectedData,
+	}, nil
+}
+
+// DeviceToResource converts an API devicepb.Device to a resource DeviceV1 and
+// assigns all default fields.
+func DeviceToResource(dev *devicepb.Device) *DeviceV1 {
```
Comment (codingllama) [inline api/types/device.go:137]: Moved from tool/tctl/common/device/resource.go, previously ProtoToResource.

File: api/types/device.go [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+			CollectedData: collectedData,
+		},
+	}
+	_ = res.CheckAndSetDefaults() // assign default fields
+	return res
+}
+
+// ResourceOSTypeToString converts OSType to a string representation suitable
+// for use in resource fields.
+func ResourceOSTypeToString(osType devicepb.OSType) string {
```
Comment (codingllama) [inline api/types/device.go:188]: All methods from here on moved from lib/devicetrust. They could be made private now, but leaving as-is.

File: api/types/device_test.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+	"google.golang.org/protobuf/testing/protocmp"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/gravitational/teleport/api/defaults"
+	devicepb "github.com/gravitational/teleport/api/gen/proto/go/teleport/devicetrust/v1"
+)
+
+// TestUnmarshalDevice tests that devices can be successfully
+// unmarshalled from YAML and JSON.
+func TestUnmarshalDevice(t *testing.T) {
```
Comment (codingllama) [inline api/types/device_test.go:33]: Moved from tool/tctl/common/device/resource_test.go. I've removed the YAML->json conversion, as it depends on a bunch of non-api imports.

**Comment by codingllama** [comment]
FYI @xacrimon @sshahcodes.

**Comment by codingllama** [comment]
CRUD follow-up: https://github.com/gravitational/teleport/pull/23913.

**Comment by codingllama** [comment]
Friendly ping @GavinFrazar @zmb3 @nklaassen ?

**Review by nklaassen — APPROVED** [review]
I'm alright with it if you think it's the way to go, but I do wonder if we could get most of the benefits of a "Resource"-shaped proto without going back under legacy/types and using gogo. I feel like we need a path away from gogo

**Review by zmb3 — APPROVED** [review]

**Pull request merged**